mm2src/coins/lp_coins.rs

```diff
@@ -3795,33 +3794,35 @@ impl DexFee {
         }
     }
 
-    /// Calculates DEX fee with a threshold based on min tx amount of the taker coin.
-    /// taker_pubkey may be optional if it is not known yet but we need total dex fee amount
-    pub fn new_from_taker_coin(
+    /// Calculates DEX fee with known taker_pubkey (for some takers dexfee may be zero).
+    pub fn new_with_taker_pubkey(
         taker_coin: &dyn MmCoin,
-        rel_ticker: &str,
+        maker_ticker: &str,
         trade_amount: &MmNumber,
-        taker_pubkey: Option<&[u8]>,
+        taker_pubkey: &[u8],
     ) -> DexFee {
-        if let Some(taker_pubkey) = taker_pubkey {
-            if !taker_coin.is_privacy() && taker_coin.burn_pubkey() == taker_pubkey {
-                return DexFee::NoFee; // no dex fee if the taker is the burn pubkey
-            }
+        if !taker_coin.is_privacy() && taker_coin.burn_pubkey() == taker_pubkey {
+            return DexFee::NoFee; // no dex fee if the taker is the burn pubkey
         }
+        Self::new_from_taker_coin(taker_coin, maker_ticker, trade_amount)
+    }
+
+    /// Calculates DEX fee with a threshold based on min tx amount of the taker coin.
+    /// With this fn we may calculate the max dex fee amount, when taker_pubkey is not known yet.
+    pub fn new_from_taker_coin(taker_coin: &dyn MmCoin, maker_ticker: &str, trade_amount: &MmNumber) -> DexFee {
         // calc dex fee
-        let rate = Self::dex_fee_rate(taker_coin.ticker(), rel_ticker);
+        let rate = Self::dex_fee_rate(taker_coin.ticker(), maker_ticker);
         let dex_fee = trade_amount * &rate;
         let min_tx_amount = MmNumber::from(taker_coin.min_tx_amount());
-        if dex_fee <= min_tx_amount {
-            return DexFee::Standard(min_tx_amount);
-        }
 
         if taker_coin.is_kmd() {
             // use a special dex fee option for kmd
             Self::calc_dex_fee_for_op_return(dex_fee, min_tx_amount)
         } else if taker_coin.should_burn_dex_fee() {
             // send part of dex fee to the 'pre-burn' account
             Self::calc_dex_fee_for_burn_account(dex_fee, min_tx_amount)
+        } else if dex_fee <= min_tx_amount {
+            DexFee::Standard(min_tx_amount)
         } else {
             DexFee::Standard(dex_fee)
         }
```

Comment:
This is a complicated condition, early returns / guard clauses will make it a lot simpler, e.g.
```rust
    if taker_coin.is_kmd() {
        return Self::calc_dex_fee_for_op_return(dex_fee, min_tx_amount);
    }
    
    if taker_coin.should_burn_dex_fee() {
        return Self::calc_dex_fee_for_burn_account(dex_fee, min_tx_amount);
    }
    
    if dex_fee <= min_tx_amount {
        return DexFee::Standard(min_tx_amount);
    }
    
    DexFee::Standard(dex_fee)
```

Reply:
fixed